### DISCUSSION TITLE
Binding Modes

### DISCUSSION THREAD

**Discussion by ghost** [YaLTeR/niri#529] [Ideas]
basically just [this](https://i3wm.org/docs/userguide.html#binding_modes).

as niri does not have it's own bar, `niri msg get_binding_mode` should exist (or similiar).

example of how config would look like:
```kdl
// "default" is implied to not break configs
binds {
    Mod+D { mode "launch"; }
    ...
}

binds "launch" {
    F { spawn firefox; mode "default"; }
    ...
    Esc { mode "default"; }
}
```

**Comment by YaLTeR** [comment]
Do you have a case where this provides a considerable improvement over just binding things to keys? I personally never found myself using these much when I was using sway, and the whole accidentally getting stuck in a mode can get rather annoying.

**Reply by ghost** [reply]
i'm using a "resize" mode in i3, i couldnt just bind those to (a different mod)+hjkl because i've run out of modifier keys

**Reply by FreeFull** [reply]
When I was using sway, I had a mode configured to disable all sway keybinds (except for the keybind to exit the mode); It's very useful when working with virtual machines, to prevent the compositor from grabbing keyboard input meant for the VM.

**Reply by YaLTeR** [reply]
That's better handled by the inhibit portal and wlr inhibit

**Reply by brainlessbitch** [reply]
> i'm using a "resize" mode in i3, i couldnt just bind those to (a different mod)+hjkl because i've run out of modifier keys

i have what i assume is a similar mode, and i as i much prefer the controlled resizing of keybinds, i really need this mode back

**Comment by ghost** [comment]
i've started implementing it

**Reply by ghost** [reply]
![image](https://github.com/YaLTeR/niri/assets/46534106/3e144071-1323-4d8e-98b9-9c8c190ca179)

**Comment by ehmzee** [comment]
Made a GitHub account to say that I would really appreciate this feature!

It'd be nice to be able to add binds more intuitively without the risk of conflicting key combinations. I primarily edit with vi-style bindings in Doom Emacs, and coming from Hyprland, it was nice being able to implement similar modal bindings throughout the rest of the compositor. I also think this could be an improvement for people who use different keyboard layouts, sizes like 60%, or for implementing binds for other, more niche peripherals (Azeron's products come to mind, for example). Being able to implement a mode for a few context-specific bindings--rather than using more complex solutions like switching the keyboard layout, relying on external programs/scripts (and preventing focus conflicts), or hardware-level configuration that may or may not be available--would be really convenient.

This has been the one feature I've sorely missed since I've recently switched to niri.

**Reply by YaLTeR** [reply]
Hmm, interesting. Do you have an example of such a setup I could look at? 

Also, I use a 40% kb myself, works just fine without binding modes :)

**Comment by Pajn** [comment]
I'd love this as well and non-modal keybinds is a big gripe for me with PaperWM. Primarily working in nvim and wezterm which both provides modal keybinds they are in my muscle memory, and I often try to do WM keybinds by tapping the modifier and then the key which of course doesn't work and I stare at the screen for a while before realizing why it didn't work.

They are also very nice for being able to bind more things without having to figure our more and more elaborate keyboard combinations. The both already mentioned launch and resize modes are something I'd definitely use.

**Reply by YaLTeR** [reply]
> I often try to do WM keybinds by tapping the modifier and then the key which of course doesn't work and I stare at the screen for a while before realizing why it didn't work.

So you want to bind a modifier by itself to go into a mode? That won't work, at least not right away, because we don't have modifier-only binds yet, and it will likely require release binds which we don't have yet either. That said, you might be able to achieve this with something like kmonad already.

**Reply by Pajn** [reply]
Ah, that's exactly what I wanted to have. But yeah, kmonad or similar could work to make a modifier tap send a chord.

But still, I'd also appreciate "normal" binds to modes to have a resize mode use hjkl and a launcher mode use simple mnemonics without worrying about collisions.

**Comment by raffomania** [comment]
Since you've asked for setups earlier in this thread, here's my old i3 setup - I'm currently looking for some way to bind these actions in niri without using too many modifiers, I'm having a hard time finding free combinations.

I hope it should be pretty self-explanatory, please ask if anything is unclear about the usage!

```
mode "notifications" {
    bindsym space exec dunstctl close-all
    bindsym j exec dunstctl close
    bindsym k exec dunstctl history-pop
    bindsym l exec dunstctl context
    bindsym Escape mode "default"; exec check-notification-count
}

mode "media" {
    bindsym h mode "default", exec playerctl previous
    bindsym l mode "default", exec playerctl next
    bindsym j mode "default", exec playerctl play-pause
    bindsym k mode "default", exec music-notify
    bindsym s mode "default", exec switch-audio
    bindsym $mod+h exec playerctl previous
    bindsym $mod+l exec playerctl next
    bindsym $mod+j exec playerctl play-pause
    bindsym $mod+k exec music-notify
    bindsym Shift+h exec playerctl position 30-
    bindsym Shift+l exec playerctl position 30+
    bindsym Shift+k exec pactl set-sink-volume @DEFAULT_SINK@ +5%
    bindsym Shift+j exec pactl set-sink-volume @DEFAULT_SINK@ -5%
    bindsym Shift+m exec pactl set-sink-mute @DEFAULT_SINK@ toggle
    bindsym Escape mode "default"
}

mode "open" {
    bindsym q mode "default", exec --no-startup-id "rofi -show calc -modi calc -no-show-match -no-sort"
    bindsym c mode "default", exec --no-startup-id "rofi -show vscode-recent -modi vscode-recent"
    bindsym d mode "default", exec thunar
    bindsym f mode "default", exec firefox
    bindsym l mode "default", exec screen_layout
    bindsym e mode "default", exec rofimoji -a clipboard
    bindsym m mode "default", exec spotify
    bindsym w mode "default", exec --no-startup-id next-wallpaper
    bindsym v mode "default", exec pavucontrol
    bindsym x mode "default", exec "alacritty -e xkill"
    bindsym m mode "default", exec "obsidian"

    bindsym --release s mode "default", exec "flameshot gui"
    bindsym p mode "default", exec "flameshot full -p $HOME"

    bindsym Escape mode "default"
}
```
